datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -989,8 +996,21 @@ impl SMJStream {
                 }
             }
             Ordering::Equal => {
-                if matches!(self.join_type, JoinType::LeftSemi) {
+                if matches!(self.join_type, JoinType::LeftSemi) && self.filter.is_some() {
+                    join_streamed = !self
+                        .streamed_batch
+                        .join_filter_matched_idxs
+                        .contains(&(self.streamed_batch.idx as u64))
+                        && !self.streamed_joined;
+                    // if the join filter specified there can be references to buffered columns
+                    // so buffered columns are needed to access them
+                    join_buffered = join_streamed;
+                }
+                if matches!(self.join_type, JoinType::LeftSemi) && self.filter.is_none() {
                     join_streamed = !self.streamed_joined;
+                    // if the join filter specified there can be references to buffered columns
```

Comment:
This block knows `self.filter.is_none()`, why you still do `join_buffered = self.filter.is_some();`?